File: src/main/java/net/botlify/anticaptchacom/HttpRequester.java
package net.botlify.anticaptchacom;

import net.botlify.anticaptchacom.exceptions.AntiCaptchaComException;
import okhttp3.*;
import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;

import java.io.IOException;

class HttpRequester {

  @NotNull
  private final OkHttpClient client;

  public HttpRequester() {
    this.client = new OkHttpClient.Builder()
        .build();
  }

  public @NotNull String sendPost(@NotNull final String url,
                                  @NotNull final JSONObject body) throws IOException, AntiCaptchaComException {
    return (HttpRequester.sendPost(client, url, body.toString()));
  }

  public @NotNull String sendPost(@NotNull final String url,
                                  @NotNull final Object body) throws IOException, AntiCaptchaComException {
    return (HttpRequester.sendPost(client, url, body));
  }

  /**
   * This method sends a POST request to the url given in parameter.
   *
   * @param url  The url.
   * @param body The body as a JSON object.
   * @return The response.
   */
  public static @NotNull String sendPost(@NotNull final OkHttpClient client,
                                         @NotNull final String url,
                                         @NotNull final Object body) throws IOException, AntiCaptchaComException {

    final Request request = new Request.Builder()
        .url(url)
        .post(RequestBody.create(body.toString(), MediaType.parse("application/json")))
        .build();
    try (final Response response = client.newCall(request).execute()) {
      final String responseBody = response.body().string();
      final int code = response.code();
      verifyRequest(code, responseBody);
      return (responseBody);
    }
  }

  /**
   * This method sends a GET request to the url given in parameter.
   *
   * @param client The HTTP client.
   * @param url    The url.
   * @return The response as String format.
   * @throws AntiCaptchaComException If the request failed.
   * @throws IOException             If the request failed.
   */
  public static @NotNull String sendGet(@NotNull final OkHttpClient client,
                                        @NotNull final String url) throws AntiCaptchaComException, IOException {
    final Request request = new Request.Builder()
        .url(url)
        .get()
        .build();
    try (final Response response = client.newCall(request).execute()) {
      final String responseBody = response.body().string();
      final int code = response.code();
      verifyRequest(code, responseBody);
      return (responseBody);
    }
  }

  /**
   * Verifies that the request was successful.
   *
   * @param code The code of the response.
   * @param body The body of the response.
   * @throws AntiCaptchaComException If the request was not successful.
   */
  public void verifyRequest(final int code,
                            @NotNull final String body) throws AntiCaptchaComException {
    // If code is between 0 and 399, the request was successful.
    if (code >= 0 && code <= 399)
      return;
    throw (new AntiCaptchaComException(code, body));
  }

}

File: src/main/kotlin/net/botlify/anticaptchacom/HttpRequester.kt
package net.botlify.anticaptchacom

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import net.botlify.anticaptchacom.enums.AntiCaptchaComError
import net.botlify.anticaptchacom.exceptions.AntiCaptchaComException
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.apache.logging.log4j.LogManager
import java.io.IOException

private const val API_URL = "https://api.anti-captcha.com"

@JsonIgnoreProperties(ignoreUnknown = true)
private class ErrorId(
    @JsonProperty("errorId")
    val errorId: AntiCaptchaComError,
)

internal class HttpRequester {

    private val client: OkHttpClient = OkHttpClient.Builder()
        .build()

    companion object {
        private val log = LogManager.getLogger(HttpRequester::class.java)
    }

    @Throws(IOException::class, AntiCaptchaComException::class)
    fun sendPost(
        path: String,
        body: String,
    ): String {
        val url = "$API_URL/$path"
        val parse = "application/json".toMediaType();
        val bodyJson = body.toRequestBody(parse);

        val request: Request = Request.Builder()
            .url(url)
            .post(bodyJson)
            .build()
        client.newCall(request).execute().use { response ->
            val responseBody = response.body.string()
            log.trace("Response POST to url $url: {}", responseBody)
            verifyRequest(responseBody)
            return (responseBody)
        }
    }

    @Throws(IOException::class, AntiCaptchaComException::class)
    fun sendGet(
        path: String,
    ): String {
        val url = "$API_URL/$path"
        val request: Request = Request.Builder()
            .url(url)
            .get()
            .build()
        client.newCall(request).execute().use { response ->
            val responseBody = response.body.string()
            log.trace("Response GET to url $url: {}", responseBody)
            verifyRequest(responseBody)
            return (responseBody)
        }
    }

    /**
     * Verifies that the request was successful.
     *
     * @param body The body of the response.
     * @throws AntiCaptchaComException If the request was not successful.
     */
    @Throws(AntiCaptchaComException::class)
    fun verifyRequest(
        body: String
    ) {
        val mapper = jacksonObjectMapper()
        val errorId = mapper.readValue(body, ErrorId::class.java).errorId
        if (errorId == AntiCaptchaComError.NO_ERROR) {
            return
        }
        val responseJson = mapper.readValue(body, AntiCaptchaComException::class.java)
        throw responseJson
    }
}
